url7b90c1010b_DastanIqbal_HelloZMQwithJava_tgz-pJ8-ClientJ8/src/org/zeromq/ZLoop.java

```diff
@@ -0,0 +1,342 @@
+    //  are specified, uses only socket. If multiple poll items exist for same
+    //  socket/FD, cancels ALL of them.
+
+    public void removePoller(PollItem item_)
+    {
+        PollItem item = item_;
+
+        Iterator<SPoller> it = pollers.iterator();
+        while (it.hasNext()) {
+            SPoller p = it.next();
+            if (item.equals(p.item)) {
+                it.remove();
+                dirty = true;
+            }
+        }
+        if (verbose)
+            System.out.printf("I: zloop: cancel %s poller (%s, %s)",
+                    item.getSocket() != null ? item.getSocket().getType() : "RAW",
+                    item.getSocket(), item.getRawSocket());
+
+    }
+
+    //  --------------------------------------------------------------------------
+    //  Register a timer that expires after some delay and repeats some number of
+    //  times. At each expiry, will call the handler, passing the arg. To
+    //  run a timer forever, use 0 times. Returns 0 if OK, -1 if there was an
+    //  error.
+
+    public int addTimer(int delay, int times, IZLoopHandler handler, Object arg)
+    {
+        STimer timer = new STimer(delay, times, handler, arg);
+
+        //  We cannot touch self->timers because we may be executing that
+        //  from inside the poll loop. So, we hold the new timer on the newTimers
+        //  list, and process that list when we're done executing timers.
+        newTimers.add(timer);
+        if (verbose)
+            System.out.printf("I: zloop: register timer delay=%d times=%d\n", delay, times);
+
+        return 0;
+    }
+
+    //  --------------------------------------------------------------------------
+    //  Cancel all timers for a specific argument (as provided in zloop_timer)
+    //  Returns 0 on success.
+
+    public int removeTimer(Object arg)
+    {
+        assert (arg != null);
+
+        //  We cannot touch self->timers because we may be executing that
+        //  from inside the poll loop. So, we hold the arg on the zombie
+        //  list, and process that list when we're done executing timers.
+        zombies.add(arg);
+        if (verbose)
+            System.out.printf("I: zloop: cancel timer\n");
+
+        return 0;
+    }
+
+    //  --------------------------------------------------------------------------
+    //  Set verbose tracing of reactor on/off
+    public void verbose(boolean verbose)
+    {
+        this.verbose = verbose;
+    }
+
+    //  --------------------------------------------------------------------------
+    //  Start the reactor. Takes control of the thread and returns when the 0MQ
+    //  context is terminated or the process is interrupted, or any event handler
+    //  returns -1. Event handlers may register new sockets and timers, and
+    //  cancel sockets. Returns 0 if interrupted, -1 if cancelled by a
+    //  handler, positive on internal error
+
+    public int start()
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 27. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.